### DISCUSSION TITLE
Markdown handler works weirdly on images

### DISCUSSION THREAD

**Discussion by EgorBron** [varabyte/kobweb#504] [Q&A] [answered]
I'm trying to style images in Markdown files.

I know, a bit better way to do this is to write some CSS or define `@InitSilk` and set the selector. But, I decided to write a markdown handler for all images, as I found something similar (but for code blocks) in [bitspittle's blog source](https://github.com/bitspittle/bitspittle.dev/blob/0bab6fe934bff0617e5e4d2d804a608c633f9c65/site/build.gradle.kts#L60).

For some reason it produces really weird code.

1. Let's say I have an empty Kobweb app with package name `net.website`.

2. I put an image to `resources/public` and created a Markdown file `HelloWorld.md` with such content:
  ```md
  # Hello world
  
  ![greeting](/untitled.png)
  ```

3. For all Markdown images I created a component:

  ```kt
  package net.website.components
  
  /* imports */
  
  @Composable
  fun MdImage(src: String, alt: String) {
      Image(src, alt, Modifier.width(50.percent))
  }
  ```

4. Then, in `site/build.gradle.kts` I added the handler:
  ```kt
  // ...
  kobweb {
      markdown {
          handlers {
              img.set { image ->
                  "net.website.components.MdImage(\"${image.destination}\", \"${image.title}\")"
              }
          }
      }
  }
  // ...
  ```

5. Finally, I ran `kobweb run`, but it throws an error:
  ```
  /* ...a huge stacktrace... */
  Found 3 error(s) and 0 warning(s). Please resolve errors to continue.

  1: file:///C:/projects/website/site/build/generated/kobweb/markdown/src/jsMain/kotlin/net/website/pages/HelloWorld.kt:15:69 Too many arguments for public fun MdImage(src: String, alt: String): Unit defined in net.website.components in file MdImage.kt
  2: file:///C:/projects/website/site/build/generated/kobweb/markdown/src/jsMain/kotlin/net/website/pages/HelloWorld.kt:17:51 @Composable invocations can only happen from the context of a @Composable function
  
  3: Execution failed for task ':site:compileKotlinJs'.
  > A failure occurred while executing org.jetbrains.kotlin.compilerRunner.GradleCompilerRunnerWithWorkers$GradleKotlinCompilerWorkAction      
     > Compilation error. See log for more details
  ```

6. Now, I look to the generated `HelloWorld.kt`:

  ```kt
  package net.website.pages
  
  import androidx.compose.runtime.*
  import com.varabyte.kobweb.core.*
  import com.varabyte.kobwebx.markdown.*
  
  @Page
  @Composable
  fun HelloWorldPage() {
      CompositionLocalProvider(LocalMarkdownContext provides MarkdownContext("HelloWorld.md", /* there should be a frontmatter */)) {
          net.website.components.ArticleLayout {
              org.jetbrains.compose.web.dom.H1(attrs = { id("hello-world") }) {
                  org.jetbrains.compose.web.dom.Text("Hello world!")
              }
              org.jetbrains.compose.web.dom.P {
                  net.website.components.MdImage("/untitled.png", "null") {
                      org.jetbrains.compose.web.dom.Text("greeting") // what?! how?
                  }
              }
          }
      }
  }
  ```
    The image title (or alt text) was moved to composable trailing lambda as `Text` composable, while title became `null`!
7. The expected output is the same code as in 6., but without trailing labmda after `MdImage` and `null` in `img.title`.

* OS: _Win_
* Kobweb libs version: _0.17.1_
* Kobweb CLI version: _0.9.15_

**Comment by EgorBron** [comment]
I also looked to Kobweb Gradle plugin, and, under the hood, [it uses `img.convention { }` and `img.children()`](https://github.com/varabyte/kobweb/blob/17e25243da037a2c3373cca33969ea97175974a4/tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/MarkdownHandlers.kt#L196). I tried this too, but nothing changed (except the `img.children()`, which is sometimes became `null`).

**Comment by bitspittle** [comment, marked as answer]
Your build.gradle.kts file should look like this:

```kotlin
kobweb {
    markdown {
        handlers {
            img.set { image ->
                val altText = image.children()
                    .filterIsInstance<Text>()
                    .map { it.literal.escapeSingleQuotedText() }
                    .joinToString("")
                this.childrenOverride = emptyList()

                """net.website.components.MdImage("${image.destination}", "$altText")"""
            }
        }
    }
}
```

I just tested that and it seems to work.

**Reply by EgorBron** [reply]
Thank you, it works! I've finally figured out that the issue was in `childrenOverride`.

**Reply by bitspittle** [reply]
You're welcome! And I'll admit my code here is not the clearest. It's not really my most elegant solution, but honestly the result of a tug of war with the markdown APIs (many of which don't even have header docs so i was figuring a lot of these issues out originally by trial and error).

Sorry it wasn't very straightforward but I'm glad you raised the question here and I'm glad you're working!
